gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```diff
@@ -88,6 +88,19 @@ static String checkInstantiable(Class<?> c) {
     return null;
   }
 
+  private static <T> T useInstanceCreator(InstanceCreator<T> instanceCreator, Type type, Class<?> rawType) {
+    T instance = instanceCreator.createInstance(type);
+    if (instance == null) {
+      throw new RuntimeException("InstanceCreator " + instanceCreator + " returned null for type " + type);
+    }
+
+    if (!rawType.isInstance(instance)) {
+      throw new ClassCastException("InstanceCreator " + instanceCreator + " created instance of wrong type;"
+          + " expected " + rawType.getName() + " but got instance of unrelated type " + instance.getClass().getName());
+    }
+    return instance;
```

Comment:
Added this for sanity, and to avoid some checks in `JsonAdapterAnnotationTypeAdapterFactory.isClassJsonAdapterFactory` to handle bad `InstanceCreator`s. Most likely Gson was already failing for bad `InstanceCreator`s before, but probably at a completely unrelated place which would have made debugging more difficult.